src/mito2/src/gc.rs

```diff
@@ -0,0 +1,420 @@
+use object_store::Entry;
+use serde::{Deserialize, Serialize};
+use snafu::{OptionExt, ResultExt as _};
+use store_api::storage::RegionId;
+use store_api::{MAX_VERSION, MIN_VERSION};
+use tokio_stream::StreamExt;
+
+use crate::access_layer::AccessLayerRef;
+use crate::error::{JoinSnafu, OpenDalSnafu, Result, UnexpectedSnafu};
+use crate::manifest::action::{RegionManifestBuilder, RegionMetaAction, RegionMetaActionList};
+use crate::manifest::manager::RegionManifestManager;
+use crate::region::ManifestContextRef;
+use crate::sst::file::{FileId, FileMeta, RegionFileId};
+use crate::sst::location;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct FileGcOption {
+    /// Lingering time before deleting files.
+    #[serde(with = "humantime_serde")]
+    pub lingering_time: Duration,
+    /// Lingering time before deleting unknown files(files with undetermine expel time).
+    /// expel time is the time when the file is considered as removed, as in removed from the manifest.
+    #[serde(with = "humantime_serde")]
+    pub unknown_file_lingering_time: Duration,
+}
+
+impl Default for FileGcOption {
+    fn default() -> Self {
+        Self {
+            lingering_time: Duration::from_secs(60 * 30), // 30 minutes
+            unknown_file_lingering_time: Duration::from_secs(60 * 60 * 24), // 1 day
+        }
+    }
+}
+pub struct LocalGcWorker {
+    pub(crate) access_layer: AccessLayerRef,
+    pub(crate) manifest_ctx: ManifestContextRef,
+    /// Lingering time before deleting files.
+    pub(crate) opt: FileGcOption,
+}
+
+impl LocalGcWorker {
+    /// concurrency of listing files per region.
+    /// This is used to limit the number of concurrent listing operations and speed up listing
+    pub const CONCURRENCY_LIST_PER_FILES: usize = 512;
+    pub async fn do_region_gc(&self) -> Result<()> {
+        // TODO(discord9): impl gc worker
+        let manifest = self.manifest_ctx.manifest().await;
+        let region_id = manifest.metadata.region_id;
+        let current_files = &manifest.files;
+
+        let recently_removed_files = self.get_removed_files_expel_times().await?;
+
+        let concurrency = current_files.len() / Self::CONCURRENCY_LIST_PER_FILES;
+        let unused_files = self
+            .list_unused_files(
+                region_id,
+                current_files.keys().cloned().collect(),
+                recently_removed_files,
+                concurrency,
+            )
+            .await?;
+
+        self.access_layer
+            .object_store()
+            .delete_iter(unused_files.iter().map(|e| e.name()))
+            .await
+            .context(OpenDalSnafu)
+    }
+
+    /// Get all the removed files in delta manifest files and their expel times.
+    /// The expel time is the time when the file is considered as removed.
+    /// Which is the last modified time of delta manifest which contains the remove action.
+    ///
+    pub async fn get_removed_files_expel_times(
```

Comment:
We can get removed files from the FilePurger or just store a list of recently deleted files in the manifest.